src/lang/base/Datatypes.ml

```diff
@@ -163,6 +163,26 @@ let scilla_list_to_ocaml v =
   in
   convert_to_list v
 
+(* Call iter for each element of the list v. *)
+let scilla_list_iterator iter v =
+  (* TODO: How to have monadic result inside a while loop? *)
+  let err = (Utils.mk_internal_error "Malformed Scilla list, error while traversing.") in
+  let seen_nil = ref false in
+  let elm = ref v in
+  while not !seen_nil do
```

Comment:
Can this not be done more elegantly with a local function:

```
let walk elm =
  match elm with
  | ADTValue ("Cons", _, _ :: ll) ->
     (iter elm;
      walk ll)
  | ADTValue ("Nil", _, _) -> ()
  | _ -> raise error
in
walk v
```

?

Reply:
Yes if course, but I was just trying to avoid recursion for anything to do with large lists. Let me see if this fixes our crash, if it dies, I'll just use it. We can use a while loop later if necessary.

Reply:
As long as it's tail recursion it should be fine.

Reply:
Fine, as in exactly the same execution-wise, just prettier to read.